### DISCUSSION TITLE
Compatible with other templates other than Grid?

### DISCUSSION THREAD

**Discussion by mkammes** [rrirower/DP2SceneGraphWizard#62] [Q&A] [answered]
Curious if anyone has used other Roku provided templates (aside from Grid) along with DP2SceneGraph.

Obviously it would be unsupported, but im not opposed to tinkering with it in an IDE to get it to work.

**Comment by rrirower** [comment]
Any template is going to be tightly married to the code. That makes it a big task to swap out one template for another. BrightScript and its xml counterpart are not all that flexible.  They are still pretty immature as coding languages.

I plan on continuing to add new features and update the design in the future. I urge any user of this project to post enhancement requests.

**Reply by mkammes** [reply]
Thanks for the input!

I've downloaded a bunch of the channel samples (including a very long-in-the-tooth hero grid template), and although I can bastardize the code in them for my nefarious purposes, the stock grid template seems to be the easiest to springboard from.

I purposely waited until the EOY to do move from DP to SDK, tracking your progress - anticipating the further you were along, the less troubleshooting I'd have to do!

re: flexible development, while I love the tools Roku provides (RALE and Stream Tester, just to name two), I'm still bummed about the extra steps for working on the UI. As of now, it's been jerky:

1. Sideload the export of DP2SceneGraph Wizard.
2. Launch RALE
3. Adjust layout of the channel, manually copying XML values if you want to move or duplicate Scenes or labels, etc. No drag and drop of existing elements.
4. Export the code (really, just copy) from RALE.
5. Open up the appropriate XML in VS Code
6. Paste the XML generated from RALE (you're on your own for syntax and troubleshooting, I believe)
7. Rezip, sideload.
8. Rinse, repeat.

I don't suppose I've overlooked something here?

Thanks again.

**Reply by rrirower** [reply, marked as answer]
Agreed.  Roku has a long way to go before they have the tools to compete with Visual Studio and its ease of use and design.

Unfortunately, you're not missing anything.  It's the nature of the beast at the moment.

**Comment by mkammes** [comment]
I’ve come up with a streamlined development environment for those who are not as code-savvy as others.

Requirements:
•	Roku Unit, with [developer mode enabled](https://developer.roku.com/fr-ca/docs/developer-program/getting-started/developer-setup.md)
•	Windows (tested) or macOS (should work)
•	[VSCode](https://code.visualstudio.com/)
•	[Roku Advanced Layout Editor](https://devtools.web.roku.com/roku-advanced-layout-editor/) (RALE)
•	ZIP file from DP2SceneGraphWizard (easier) or [Roku Sample templates from GitHub](https://github.com/rokudev/samples) (more difficult)


1.	Unzip the DP2SceneGraph ZIP into a local folder.  Let’s call it “workarea”. It should contain the components, images, and source folders, plus your manifest file.
a.	If you know what you’re doing, you can take a Roku Sample template from GitHub and place it in this folder instead; but you’ll need to do a bit more code work, including the manifest file. _This is outside the scope of this workflow overview._
2.	Open RALE. Connect to your Roku.
3.	In RALE, click “Select Channel”, then “Directory”, not “ZIP”.  Point to the “workarea” folder, and click the upload button. Your channel will be sideloaded, and your RALE interface should show your Roku channel UI.
4.	Open VSCode. Open the “workarea” folder. 
5.	Manipulate the values in RALE until you get what you like. Because RALE is limited to what XMLs it will output (_“Default nodes could not be exported. Only created nodes that were added by RALE will be exported.”_), you need to manually make the code changes in VSCode. **I know, this sucks.**
6.	Save the changed file(s) in VSCode.
7.	In RALE hit the “Refresh” circle icon. This will force the code to be re-sideloaded into the Roku unit. Because you have changed the files in VSCode, the updated code is what will be pushed to your Roku.
8.	Rinse, repeat.
9.	Bonus: Save the “workarea” folder as a “workspace” in VSCode. If you have CoPilot, you now can ask CoPilot about your code and even have it generate updated code. This has saved me hours already!
